File: dvs_.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Wed Mar 23 12:25:47 2022

@author: marcos
"""

# Assess WOFOST simulations

import matplotlib
#from matplotlib.ticker import FormatStrFormatter
import matplotlib.pyplot as plt
import numpy as np
from sys import exit
#import matplotlib.gridspec as gridspec
import time as tm

def locvar(varname,lines):
#  varname: name of the variable given in the file
#  lines  : list of lines returned by readlines() procedure    
   c1 = [] 
   for l in lines:
        l = l.split("=")
        c1.append(l[0].strip().lower())
   return c1.index(varname.lower())     




def RLR(varname,lines):
#  Purpose: reads a list type variable anywhere in a file and assigns as real
# Example of reading:
#    DTSMTB   =    0.00,    0.00,     ! daily increase in temp. sum
#                  35.00,   1.00,     ! as function of av. temp. [cel; cel d]
#                  45.00,   0.5
#    DVSI     =    0.    ! initial DVS
#  Argumments:
#  varname: name of the variable given in the file
#  lines  : list of lines returned by readlines() procedure    

    loc = locvar(varname, lines)
    
    lOut = []
    ldummy = lines[loc].split("=")
    ldummy = ldummy[1].split(",")
    lOut.append(float(ldummy[0]))
    lOut.append(float(ldummy[1]))
#    print(lOut)
    
    flag = True
    while(flag):
         loc = loc + 1
         ldummy = lines[loc].split(",")
         lOut.append(float(ldummy[0]))
         lOut.append(float(ldummy[1]))
         flag = not("=" in lines[loc+1]) and "," in lines[loc]
#         print(flag)
    return np.array(lOut)

def readchar(varname,lines):
#  Argumments:
#  varname: name of the variable given in the file
#  lines  : list of lines returned by readlines() procedure    
  loc = locvar(varname, lines)
  l = lines[loc].split("=")
  if ("!" in l[1]):
      ll = l[1].split("!")
      vout = ll[0].strip()
      if ((vout[0] == '"' and not(vout[-1] == '"')) or \
             (vout[0] == "'" and not(vout[-1]) == "'")):
          print("Error in char value format for " + varname)
          exit()
      elif(vout[0] == '"' or vout[0] == "'"):
          return vout[1:-1]
      else:
          return vout 
  else:
      vout = l[1].strip()
      if ((vout[0] == '"' and not(vout[-1] == '"')) or \
             (vout[0] == "'" and not(vout[-1]) == "'")):
          print("Error in char value format for " + varname)
          exit()
      elif(vout[0] == '"' or vout[0] == "'"):
          return vout[1:-1]
      else:
          return vout 
  
def readfloat(varname,lines):
#  Argumments:
#  varname: name of the variable given in the file
#  lines  : list of lines returned by readlines() procedure    
  loc = locvar(varname, lines)
  l = lines[loc].split("=")
  if ("!" in l[1]):
      ll = l[1].split("!")
      vout = ll[0].strip()
      return float(vout)
  else:
      vout = l[1].strip()
      return float(vout)
  
    
    

def afgen(ldummy,x):
# Linear interpolation fuction
    if (x < ldummy[0]):
        return ldummy[1]
    if (x > ldummy[-2]):
        return ldummy[-1]     
    for i in np.arange(2,len(ldummy)):
        if (x <= ldummy[i]):
            slope = (ldummy[i+1]-ldummy[i-1])/(ldummy[i]-ldummy[i-2])
            return ldummy[i-1] + slope * (x - ldummy[i-2])
       
#def ft1900(dd,mes,ano,hora,mm,ss):
def strtoFloat(strd1):
   if (strd1 == ""):
       return -999
   else:
       return float(strd1)       

def datefmt(stVar):
# stVar is a strutc_time var and returns in a string format
   return tm.strftime("%d-%m-%Y",stVar)
    


# ******************** Start program ***********************************
cropfile = "cassava.w41" 
meteofile = "dados_meteo.csv" 

print("Before starting, verify if the following files are in the same\
directoy of this program:")
print(cropfile)
print(meteofile)
input("Press Enter to continue")

f = open(cropfile,"r")
lines = f.readlines()
f.close()

# Reading some variables from wofost crop file
dtsmtb = RLR("dtsmtb",lines)
plantdateST = tm.strptime(readchar("dateplant",lines),"%d-%m-%Y")
tbemerg = readfloat("TBASEM", lines)
TsumEm  = readfloat("TSUMEM",lines)
tsum1   = readfloat("tsum1", lines)
tsum2   = readfloat("tsum2", lines)



# Read meteo file

f = open(meteofile,"r")
lnmeteo = f.readlines()
f.close()


# Variables in the meteofile
dateST,Tmed,Tmin,Tmax,Urmed,Urmin,Urmax,Vv2m,Hg,P,ETo,ETc = \
    [],[],[],[],[],[],[],[],[],[],[],[]


#print(lnmeteo[0])
for l in lnmeteo[2:]:
    l = l.split(",")
    idatestr = l[0].strip()+"-"+ l[1].strip()+"-" + l[2].strip()
    dateST.append(tm.strptime(idatestr,"%d-%m-%Y"))
    Tmed.append(strtoFloat(l[3]))
    Tmin.append(strtoFloat(l[4]))
    Tmax.append(strtoFloat(l[5]))
    Urmed.append(strtoFloat(l[6]))
    Urmin.append(strtoFloat(l[7]))
    Urmax.append(strtoFloat(l[8]))
    Vv2m.append(strtoFloat(l[9]))
    Hg.append(strtoFloat(l[10]))
    P.append(strtoFloat(l[11]))
    ETo.append(strtoFloat(l[12]))
    ETc.append(strtoFloat(l[13]))

if (Tmed.count(-999) > 0):
    print("Missing temperature values. \n It will be necessary a function for\
          for missing data")
    exit()



iPdate = 0 ; flag = True
while (flag):
    if dateST[iPdate] >= plantdateST:
        flag = False
        iPdate = iPdate - 1
    iPdate = iPdate + 1

# ipdate: position in general meteo file for the plant date

# **** Generating file with DVS calculated as function of DATE and DAP *****

#   openning file 

f1 = open("dvs.out","w")
f1.write("*DSV program by Marcos Alex\n")
f1.write("*Find below the values of dvs calculated with information taken from\
 the following files\n")
f1.write("*meteorological file name : " + meteofile + "\n")
f1.write("*crop file name           : " + cropfile  + "\n")
f1.write(30*"*"+"\n")
f1.write('{:15} {:6} {:8}\n'.format("*Date","DAP","DVS"))
f1.write(30*"*"+"\n")

pEm = 0; idvs = 0
iemtsum = 0.0; flag = True

i = iPdate ; dap = 0
while (flag):
    idate = dateST[i]
    itmed = Tmed[i]
# determine firts the day of emergency
    iemtsum = iemtsum + (itmed - tbemerg)

    if(iemtsum < TsumEm):
       idvs = 0
       pEm = pEm + 1
    else:
       tav = afgen(dtsmtb, itmed)
       if (idvs <= 1.0):
           dvr = tav/tsum1
       elif(idvs < 2.0):
           dvr = tav/tsum2
       else:
           flag = False
           
       idvs = idvs + dvr
    i = i + 1
    f1.write('{:12} {:6} {:6.2f}\n'.format(datefmt(idate),dap,idvs) )
    dap = dap + 1    
f1.close()               
     
print("")  
print("OK")
print("file dvs.out was created")
     

# PEm is the first position in vec when the plant emerges

    
    

# t1970 = tm.time()
#local_time = tm.ctime(t1970)
# localtimeTM = tm.localtime(t1970) # It store in strutc_time format
# This variable also returns the Julian day

#print(localtimeTM)
#print("\nYear :", localtimeTM.tm_year)
#print("tm_hour : ", localtimeTM.tm_hour)
#print("Julian Day :", localtimeTM.tm_yday)

# time.strftime takes an epoch time in strutc_time format and returns in str format
# according to your  formatting settings
#timeString = tm.strftime("%d/%m/\n%Y",localtimeTM)
#print(timeString)

# strptime() takes a string and corresponding time format to return strutc_time 
# since epoch
# Example:
    
# date1 = "24-Mar-2022"
# date2 = "25-Mar-2022"
# date1ST = tm.strptime(date1,"%d-%b-%Y")
# date2ST = tm.strptime(date2,"%d-%b-%Y")
# print(date1ST)
# print(date2ST > date1ST)



# Convert strutc_time in float epoch

    
    

File: dvs.f90
program read_files_python
implicit none

integer,parameter                :: maxL = 100
integer,parameter                :: maxL2 = 1000
integer,parameter                :: maxD = 1000

real*8,allocatable,dimension(:)  :: Tmed
real*8,allocatable,dimension(:)  :: t1900
integer,allocatable,dimension(:) :: dia,mes,ano
character(len=10),allocatable, &
          &dimension(:)          :: date



character(len=40)               :: dir_file
real*8,allocatable,dimension(:) :: dtsmtb  
real*8,dimension(maxL)          :: dummy1
real*8,dimension(maxL2)         :: dummy2

real*8                          :: tbasem
real*8                          :: tsumem
real*8                          :: tsum1
real*8                          :: tsum2
real*8                          :: iemtsum
real*8                          :: dateplant1900
real*8                          :: tav
real*8                          :: dvr
real*8                          :: idvs

character(len=10)               :: dateplant
integer                         :: i
integer                         :: iplant
integer                         :: dap = 0


print*,"Before starting, verify if the following files are in the same &
       & directoy of this program:"
print*,"cassava.w41"
print*,"dados_meteo.csv"
print*,""
print*,"Press Enter to continue"
read(*,*)




! ---- reading crop file ----------------------

dir_file = "cassava.w41" 
call readcropfile(dir_file)



! read meteofile
call readmeteodata("dados_meteo.csv")


! Startting the main computations

open(10,file="dvs_fortran.out",status="replace")

write(10,"(A)")"*DSV program by Marcos Alex"
write(10,"(A)")"*Find below the values of dvs calculated with information taken from the following files"
write(10,"(A)")"*meteorological file name : dados_meteo.csv"
write(10,"(A)")"*crop file name           : cassava.w41"

write(10,"(A)")"****************************************"
write(10,"(A12,A5,A10)")"Data","DAP","DVS"
write(10,"(A)")"****************************************"

i = 1
do while(t1900(i) < dateplant1900  )
 i = i + 1
enddo

iemtsum = 0.d0
idvs = 0.d0
do while (idvs < 2.0 .and. i < size(Tmed)) 

  iemtsum = iemtsum + (Tmed(i) - tbasem)
  if (iemtsum < tsumem) then
     idvs = 0d0
  else
     tav = afgen(dtsmtb,Tmed(i))
     if (idvs <= 1.0) then
         dvr = tav/tsum1
     else
         dvr = tav/tsum2
     endif
     
     idvs = idvs + dvr
  endif
  dap = dap + 1
  print*,date(i+1)
  write(10,"(A12,I5,F10.3 )")date(i+1),dap,idvs
  i = i + 1
enddo


close(10)

print*,""
print*,"OK"
print*,"file dvs_fortran.out was created"
print*,""
print*,"Press Enter to close the program"
read(*,*)



contains 

subroutine readcropfile(filename)
implicit none
character(len=*) :: filename
integer         :: ilc
integer,dimension(6):: datea



call RDINIT (30,40,filename)

call RDADOU ("dtsmtb",dummy1,maxL,ilc)
allocate(dtsmtb(ilc))
dtsmtb = dummy1(1:ilc)

call RDSDOU ("TBASEM",tbasem)
call RDSDOU ("tsum1",tsum1)
call RDSDOU ("tsum2",tsum2)
call RDSDOU ("tsumem",tsumem)
call RDSCHA ("dateplant",dateplant)

! convert dateplant to 1900 time
read(dateplant(7:10),"(I4)")  datea(1)
read(dateplant(4:5),"(I2)")   datea(2)
read(dateplant(1:2),"(I2)")   datea(3)
do i = 4,6
  datea(i) = 0
enddo
call DTARDP(datea,0.0,dateplant1900)
!----------------------------------------




close(30)
call RDDTMP(10)

end subroutine readcropfile


subroutine chartoint(ch,varint)
implicit none
character(len=*),intent(in)   :: ch
integer,intent(out)           :: varint

if (ch == "") then
   print*,"Warning: there are missing data in the meteofile"
   varint = -999
else
  read(ch,'(I10)')varint
endif

end subroutine chartoint

subroutine chartoR8(ch,varout)
implicit none
character(len=*),intent(in)   :: ch
real*8,intent(out)            :: varout

if (ch == " ") then
   print*,"Warning: there are missing data in the meteofile"
   varout = -999.9D0
else
  read(ch,'(F10.5)')varout
endif

end subroutine chartoR8

subroutine sepchar(char1,varch,fildsep)
implicit none
character(len=*),intent(in):: fildsep
character(len=*),intent(in):: char1
character(len=*),dimension(4),intent(out):: varch

integer n1,i
character(len = 5)              ::ch

 n1 = 0
 ch = ''
 do i=1,len(char1)
   if(char1(i:i) == fildsep .and. n1 <=4) then
      n1 = n1 + 1
      varch(n1) = ch
      ch =""
   else
      ch = trim(ch)//char1(i:i)
   endif
 enddo
end subroutine sepchar

! -------------------------------------------
function datefmtFchar(dd,mm,yyyy)
implicit none
character(len=10)          :: datefmtFchar
character(len=*),intent(in)::dd,mm,yyyy
character(len=2)           :: idd,imm

if (len(trim(dd)) == 1) then
  idd = "0"//trim(dd)
else
  idd = trim(dd)
endif

if (len(trim(mm)) == 1) then
   imm = "0"//trim(mm)
else
  imm = trim(mm)
endif

datefmtFchar = idd//"-"//imm//"-"//trim(yyyy)


end function datefmtFchar

subroutine readmeteodata(fn)
implicit none 
character(len=*),intent(in):: fn
! locals
integer,allocatable,dimension(:) :: id1,id2,id3
real*8,allocatable,dimension(:)  :: rd1
character(len=10),allocatable, &
          &dimension(:)          :: dumdate
character(len=10)                :: itimefrm
character(len=5),dimension(4)    :: varch
character(len=16)                :: char1
real*8,allocatable,dimension(:)  :: t1900D
integer,dimension(6)             :: datea

integer                          :: IO = 0
integer                          :: n,i

open(10,file=trim(fn),status="old")

! skip two lines
read(10,*) ; read(10,*)

! Allocating  dummy 1-D arrays
allocate(id1(maxD))
allocate(id2(maxD))
allocate(id3(maxD))
allocate(rd1(maxD))
allocate(dumdate(maxD))
allocate(t1900D(maxD))

n = 0
do while (IO == 0)
 read(10,"(A16)",iostat= io)char1
 call sepchar(char1,varch,",")   ! Gets a char varriable, separete by "," and stores into 
                                 ! char1 character 1-D array
 n = n + 1
 dumdate(n) = datefmtFchar(varch(1),varch(2),varch(3))

 call chartoint(varch(1),id1(n))
 call chartoint(varch(2),id2(n))
 call chartoint(varch(3),id3(n))
 call chartoR8(varch(4),rd1(n))

!converting date to t1900 real time
 datea(1) = id3(n)
 datea(2) = id2(n)
 datea(3) = id1(n)
 datea(4) = 0; datea(5) = 0; datea(6) = 0
 call DTARDP(datea,0.0,t1900D(n))
 
enddo

!-- Allocating the variables day,month and year ,and Tmed
allocate(dia(n))
allocate(mes(n))
allocate(ano(n))
allocate(Tmed(n))
allocate(date(n))
allocate(t1900(n))

dia  = id1(1:n)
mes  = id2(1:n)
ano  = id3(1:n)
Tmed = rd1(1:n)
date = dumdate(1:n)
t1900 = t1900D(1:n)

! Deallocating the dummy 1d arrays
deallocate(id1);deallocate(id2);deallocate(id3);deallocate(rd1)
deallocate(dumdate); deallocate(t1900D)
close(10)

end subroutine readmeteodata


function afgen(array,x)
! Interpolates linearly an 1-D array values in the format x1,y1,x2,y2,...xn,yn
implicit none
real*8 afgen
real*8,dimension(:),intent(in) :: array
real*8,intent(in)              :: x

! locals
real*8  :: slope
integer :: i,n

n = size(array)

if ( x < array(1)) then
    afgen = array(2)
    return
else
   do i = 3,n-1,2
      if ( x <= array(i)) then
         slope = (array(i+1) - array(i-1))/(array(i) - array(i-2))
         afgen = array(i-1) + slope * (x - array(i-2))
         return
      endif
   enddo
endif 

! It means x is greater than last x in the array
afgen = array(n)

end function afgen


end program read_files_python
